core/src/main/java/io/grpc/internal/ClientCallImpl.java

```diff
@@ -333,46 +340,97 @@ private static void logIfContextNarrowedTimeout(
 
   private void removeContextListenerAndCancelDeadlineFuture() {
     context.removeListener(cancellationListener);
-    ScheduledFuture<?> f = deadlineCancellationFuture;
+    ScheduledFuture<?> f = deadlineCancellationSendToServerFuture;
+    if (f != null) {
+      f.cancel(false);
+    }
+
+    f = deadlineCancellationNotifyApplicationFuture;
     if (f != null) {
       f.cancel(false);
     }
   }
 
-  private class DeadlineTimer implements Runnable {
-    private final long remainingNanos;
+  private ScheduledFuture<?> startDeadlineNotifyApplicationTimer(Deadline deadline,
+      final Listener<RespT> observer) {
+    final long remainingNanos = deadline.timeRemaining(TimeUnit.NANOSECONDS);
 
-    DeadlineTimer(long remainingNanos) {
-      this.remainingNanos = remainingNanos;
+    class DeadlineExceededNotifyApplicationTimer implements Runnable {
+      @Override
+      public void run() {
+        Status status = buildDeadlineExceededStatusWithRemainingNanos(remainingNanos);
+        delayedCancelOnDeadlineExceeded(status, observer);
+      }
     }
 
-    @Override
-    public void run() {
-      InsightBuilder insight = new InsightBuilder();
-      stream.appendTimeoutInsight(insight);
-      // DelayedStream.cancel() is safe to call from a thread that is different from where the
-      // stream is created.
-      long seconds = Math.abs(remainingNanos) / TimeUnit.SECONDS.toNanos(1);
-      long nanos = Math.abs(remainingNanos) % TimeUnit.SECONDS.toNanos(1);
-
-      StringBuilder buf = new StringBuilder();
-      buf.append("deadline exceeded after ");
-      if (remainingNanos < 0) {
-        buf.append('-');
+    return deadlineCancellationExecutor.schedule(
+        new LogExceptionRunnable(new DeadlineExceededNotifyApplicationTimer()),
+        remainingNanos,
+        TimeUnit.NANOSECONDS);
+  }
+
+  private Status buildDeadlineExceededStatusWithRemainingNanos(long remainingNanos) {
```

Comment:
nit: this method is used only once, and the only caller is very short. It may be more readable as an inline codeblock at the call site.